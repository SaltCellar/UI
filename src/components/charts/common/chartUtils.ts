import { Report } from 'api/reports/report';
import endOfMonth from 'date-fns/end_of_month';
import format from 'date-fns/format';
import formatDate from 'date-fns/format';
import getDate from 'date-fns/get_date';
import getYear from 'date-fns/get_year';
import startOfMonth from 'date-fns/start_of_month';
import {
  ComputedReportItem,
  getComputedReportItems,
} from 'utils/computedReport/getComputedReportItems';
import {
  FormatOptions,
  unitLookupKey,
  ValueFormatter,
} from 'utils/formatValue';
import { SortDirection } from 'utils/sort';
import { t } from '../../i18nProvider';

export interface ChartDatum {
  childName?: string;
  key: string | number;
  name?: string | number;
  show?: boolean;
  tooltip?: string;
  units: string;
  x: string | number;
  y: number;
}

// The computed report cost or usage item
export const enum ComputedReportItemType {
  cost = 'cost', // cost.total.value
  infrastructure = 'infrastructure', // infrastructure.total.value
  supplementary = 'supplementary', // supplementary.total.value
  usage = 'usage', // usage.value
}

// The computed report value
export const enum ComputedReportItemValueType {
  none = 'none', // A value type is not used in this scenario (e.g., usage.value)
  markup = 'markup', // infrastructure.markup.value
  raw = 'raw', // infrastructure.raw.value
  total = 'total', // // infrastructure.total.value
  usage = 'usage', // infrastructure.usage.value
}

export const enum ChartType {
  rolling,
  daily,
  monthly,
}

export function transformReport(
  report: Report,
  type: ChartType = ChartType.daily,
  key: any = 'date',
  reportItem: string = 'cost',
  reportItemValue: string = 'total'
): ChartDatum[] {
  if (!report) {
    return [];
  }
  const items = {
    idKey: key,
    report,
    reportItemValue,
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = getComputedReportItems(items);
  if (type === ChartType.daily) {
    return computedItems.map(i =>
      createDatum(i[reportItem], i, key, reportItem)
    );
  }
  if (type === ChartType.monthly) {
    return computedItems.map(i =>
      createDatum(i[reportItem], i, key, reportItem)
    );
  }
  return computedItems.reduce<ChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    return [...acc, createDatum(prevValue + d[reportItem], d, key, reportItem)];
  }, []);
}

export function createDatum<T extends ComputedReportItem>(
  value: number,
  computedItem: T,
  idKey = 'date',
  reportItem: string = 'cost'
): ChartDatum {
  const xVal = idKey === 'date' ? getDate(computedItem.id) : computedItem.label;
  const yVal = isFloat(value)
    ? parseFloat(value.toFixed(2))
    : isInt(value)
    ? value
    : 0;
  return {
    x: xVal,
    y: yVal,
    key: computedItem.id,
    name: computedItem.id,
    units: computedItem.units[reportItem],
  };
}

export function getDatumDateRange(
  datums: ChartDatum[],
  offset: number = 0
): [Date, Date] {
  if (!(datums && datums.length)) {
    const today = new Date();

    // If datums is empty, obtain the month based on offset (e.g., to show previous month in chart legends)
    if (offset) {
      today.setMonth(today.getMonth() - offset);
    }
    const firstOfMonth = startOfMonth(today);
    return [firstOfMonth, today];
  }

  const start = new Date(datums[0].key + 'T00:00:00');
  const end = new Date(datums[datums.length - 1].key + 'T00:00:00');
  return [start, end];
}

export function getDateRange(
  datums: ChartDatum[],
  firstOfMonth: boolean = true,
  lastOfMonth: boolean = false,
  offset: number = 0
): [Date, Date] {
  const [start, end] = getDatumDateRange(datums, offset);

  // Show the date range we are trying to cover (i.e., days 1-30/31)
  if (firstOfMonth && start.setDate) {
    start.setDate(1);
  }
  if (lastOfMonth && start.setDate) {
    const lastDate = endOfMonth(start).getDate();
    end.setDate(lastDate);
  }
  return [start, end];
}

export function getDateRangeString(
  datums: ChartDatum[],
  firstOfMonth: boolean = false,
  lastOfMonth: boolean = false,
  offset: number = 0
) {
  const [start, end] = getDateRange(datums, firstOfMonth, lastOfMonth, offset);

  return t(`chart.date_range`, {
    count: getDate(end),
    endDate: formatDate(end, 'DD'),
    month: Number(formatDate(start, 'M')) - 1,
    startDate: formatDate(start, 'DD'),
    year: getYear(end),
  });
}

export function getMonthRangeString(
  datums: ChartDatum[],
  key: string = 'chart.month_legend_label',
  offset: number = 0
): [string, string] {
  const [start, end] = getDateRange(datums, true, false, offset);

  return [
    t(key, {
      month: Number(formatDate(start, 'M')) - 1,
    }),
    t(key, {
      month: Number(formatDate(end, 'M')) - 1,
    }),
  ];
}

export function getMaxValue(datums: ChartDatum[]) {
  let max = 0;
  if (datums && datums.length) {
    datums.forEach(datum => {
      if (datum.y > max) {
        max = datum.y;
      }
    });
  }
  return max;
}

export function getTooltipContent(formatValue) {
  return function labelFormatter(
    value: number,
    unit: string = null,
    options: FormatOptions = {}
  ) {
    const lookup = unitLookupKey(unit);
    switch (lookup) {
      case 'core-hours':
      case 'hrs':
      case 'gb':
      case 'gb-hours':
      case 'gb-mo':
      case 'vm-hours':
        return t(`unit_tooltips.${lookup}`, {
          value: `${formatValue(value, unit, options)}`,
        });
      default:
        return `${formatValue(value, unit, options)}`;
    }
  };
}

export function getTooltipLabel(
  datum: ChartDatum,
  formatValue: ValueFormatter,
  formatOptions?: FormatOptions,
  idKey: any = 'date',
  units?: string
) {
  if (!datum.key) {
    return '';
  }
  if (idKey === 'date') {
    const date = format(datum.key, 'DD MMM YYYY');
    return `${date} ${formatValue(
      datum.y,
      units ? units : datum.units,
      formatOptions
    )}`;
  }
  return datum.key.toString();
}

export function getCostRangeString(
  datums: ChartDatum[],
  key: string = 'chart.cost_legend_label',
  firstOfMonth: boolean = false,
  lastOfMonth: boolean = false,
  offset: number = 0
) {
  const [start, end] = getDateRange(datums, firstOfMonth, lastOfMonth, offset);

  return t(key, {
    count: getDate(end),
    endDate: formatDate(end, 'D'),
    month: Number(formatDate(start, 'M')) - 1,
    startDate: formatDate(start, 'D'),
    year: getYear(end),
  });
}

export function getUsageRangeString(
  datums: ChartDatum[],
  key: string = 'chart.usage_legend_label',
  firstOfMonth: boolean = false,
  lastOfMonth: boolean = false,
  offset: number = 0
) {
  return getCostRangeString(datums, key, firstOfMonth, lastOfMonth, offset);
}

function isInt(n) {
  return Number(n) === n && n % 1 === 0;
}

function isFloat(n) {
  return Number(n) === n && n % 1 !== 0;
}
