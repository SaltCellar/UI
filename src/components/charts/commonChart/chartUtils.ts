import { AwsReport } from 'api/awsReports';
import { OcpReport } from 'api/ocpReports';
import endOfMonth from 'date-fns/end_of_month';
import format from 'date-fns/format';
import getDate from 'date-fns/get_date';
import startOfMonth from 'date-fns/start_of_month';
import i18next from 'i18next';
import {
  FormatOptions,
  unitLookupKey,
  ValueFormatter,
} from 'utils/formatValue';
import {
  ComputedAwsReportItem,
  getComputedAwsReportItems,
} from 'utils/getComputedAwsReportItems';
import {
  ComputedOcpReportItem,
  getComputedOcpReportItems,
} from 'utils/getComputedOcpReportItems';
import { SortDirection } from 'utils/sort';

export interface ChartDatum {
  x: string | number;
  y: number;
  key: string | number;
  name?: string | number;
  show?: boolean;
  units: string;
}

export const enum ChartType {
  rolling,
  daily,
  monthly,
}

export function transformAwsReport(
  report: AwsReport,
  type: ChartType = ChartType.daily,
  key: any = 'date',
  reportItem: any = 'total'
): ChartDatum[] {
  if (!report) {
    return [];
  }
  const items = {
    report,
    idKey: key,
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = getComputedAwsReportItems(items);
  if (type === ChartType.daily) {
    return computedItems.map(i => createDatum(i[reportItem], i, key));
  }
  if (type === ChartType.monthly) {
    return computedItems.map(i => createDatum(i[reportItem], i, key));
  }

  return computedItems.reduce<ChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    return [...acc, createDatum(prevValue + d[reportItem], d, key)];
  }, []);
}

export function transformOcpReport(
  report: OcpReport,
  type: ChartType = ChartType.daily,
  key: any = 'date',
  reportItem: any = 'cost'
): ChartDatum[] {
  if (!report) {
    return [];
  }
  const items = {
    report,
    idKey: key,
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = getComputedOcpReportItems(items);

  if (type === ChartType.daily) {
    return computedItems.map(i => createDatum(i[reportItem], i, key));
  }
  if (type === ChartType.monthly) {
    return computedItems.map(i => createDatum(i[reportItem], i, key));
  }

  return computedItems.reduce<ChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    return [...acc, createDatum(prevValue + d[reportItem], d, key)];
  }, []);
}

export function createDatum(
  value: number,
  computedItem: ComputedAwsReportItem | ComputedOcpReportItem,
  idKey = 'date'
): ChartDatum {
  const xVal = idKey === 'date' ? getDate(computedItem.id) : computedItem.label;
  const yVal = isFloat(value)
    ? parseFloat(value.toFixed(2))
    : isInt(value)
    ? value
    : 0;
  return {
    x: xVal,
    y: yVal,
    key: computedItem.id,
    name: computedItem.id,
    units: computedItem.units,
  };
}

export function getDatumDateRange(datums: ChartDatum[]): [Date, Date] {
  if (!(datums && datums.length)) {
    const today = new Date();
    const firstOfMonth = startOfMonth(today);
    return [firstOfMonth, today];
  }

  const start = new Date(datums[0].key + 'T00:00:00');
  const end = new Date(datums[datums.length - 1].key + 'T00:00:00');
  return [start, end];
}

export function getDateRange(
  datums: ChartDatum[],
  firstOfMonth: boolean = true,
  lastOfMonth: boolean = false
): [Date, Date] {
  const [start, end] = getDatumDateRange(datums);

  // Show the date range we are trying to cover (i.e., days 1-30/31)
  if (firstOfMonth && start.setDate) {
    start.setDate(1);
  }
  if (lastOfMonth && start.setDate) {
    const lastDate = endOfMonth(start).getDate();
    end.setDate(lastDate);
  }
  return [start, end];
}

export function getDateRangeString(
  datums: ChartDatum[],
  firstOfMonth: boolean = false,
  lastOfMonth: boolean = false,
  current: boolean = false
) {
  const [start, end] = getDateRange(datums, firstOfMonth, lastOfMonth);

  const monthName = format(start, 'MMM');
  const startDate = getDate(start);
  const endDate = getDate(end);

  if (current) {
    return i18next.t(`date.range_current`, {
      date: getDateString(startDate),
      month: startDate !== endDate ? monthName : '',
    });
  }
  return i18next.t(`date.range_full`, {
    endDate: getDateString(endDate),
    startDate: getDateString(startDate),
    month: startDate !== endDate ? monthName : '',
  });
}

export function getMaxValue(datums: ChartDatum[]) {
  let max = 0;
  if (datums && datums.length) {
    datums.forEach(datum => {
      if (datum.y > max) {
        max = datum.y;
      }
    });
  }
  return max;
}

export function getTooltipContent(formatValue) {
  return function labelFormatter(
    value: number,
    unit: string = null,
    options: FormatOptions = {}
  ) {
    const lookup = unitLookupKey(unit);
    switch (lookup) {
      case 'hrs':
      case 'gb':
        return i18next.t(`units.${lookup}`, {
          value: `${formatValue(value, unit, options)}`,
        });
      default:
        return `${formatValue(value, unit, options)}`;
    }
  };
}

export function getTooltipLabel(
  datum: ChartDatum,
  formatValue: ValueFormatter,
  formatOptions?: FormatOptions,
  idKey: any = 'date'
) {
  if (!datum.key) {
    return '';
  }
  if (idKey === 'date') {
    const date = format(datum.key, 'MMM D YYYY');
    return `${date}: ${formatValue(datum.y, datum.units, formatOptions)}`;
  }
  return datum.key.toString();
}

function isInt(n) {
  return Number(n) === n && n % 1 === 0;
}

function isFloat(n) {
  return Number(n) === n && n % 1 !== 0;
}

function getDateString(date) {
  const day = date % 10;
  if (day === 1) {
    return i18next.t(`date.first`, { date });
  } else if (day === 2) {
    return i18next.t(`date.second`, { date });
  } else if (day === 3) {
    return i18next.t(`date.third`, { date });
  }
  return i18next.t(`date.tenth`, { date });
}
