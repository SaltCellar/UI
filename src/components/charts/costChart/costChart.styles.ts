import { StyleSheet } from '@patternfly/react-styles';
import {
  global_disabled_color_100,
  global_disabled_color_200,
  global_FontFamily_sans_serif,
  global_FontSize_md,
  global_spacer_lg,
} from '@patternfly/react-tokens';
import { VictoryStyleInterface } from 'victory';

export const chartStyles = {
  currentInfrastructureData: {
    data: {
      fill: 'none',
      stroke: '#88D080',
      strokeDasharray: '3,3',
    },
  } as VictoryStyleInterface,
  currentUsageData: {
    data: {
      fill: 'none',
      stroke: '#A2DA9C',
    },
  } as VictoryStyleInterface,
  legend: {
    labels: {
      fontFamily: global_FontFamily_sans_serif.value,
      fontSize: 14,
    },
    minWidth: 175,
  },
  previousInfrastructureData: {
    data: {
      fill: 'none',
      stroke: global_disabled_color_200.value,
      strokeDasharray: '3,3',
    },
  } as VictoryStyleInterface,
  previousUsageData: {
    data: {
      fill: 'none',
      stroke: global_disabled_color_200.value,
    },
  } as VictoryStyleInterface,
  // See: https://github.com/project-koku/koku-ui/issues/241
  currentColorScale: ['#A2DA9C', '#88D080', '#6EC664', '#519149', '#3C6C37'],
  // TBD: No grey scale, yet
  previousColorScale: [
    global_disabled_color_200.value,
    global_disabled_color_100.value,
  ],
  yAxis: {
    axisLabel: {
      padding: 15,
    },
    grid: {
      stroke: 'none',
    },
    ticks: {
      stroke: 'none',
    },
    tickLabels: {
      fontSize: 0,
    },
  } as VictoryStyleInterface,
  xAxis: {
    axisLabel: {
      padding: 15,
    },
    grid: {
      stroke: 'none',
    },
    ticks: {
      stroke: 'none',
    },
  } as VictoryStyleInterface,
};

export const styles = StyleSheet.create({
  chartContainer: {
    ':not(foo) svg': {
      overflow: 'visible',
    },
    marginTop: global_spacer_lg.value,
  },
  chartTitle: {
    fontSize: global_FontSize_md.value,
    marginTop: global_spacer_lg.value,
  },
  legend: {
    display: 'inline-block',
    fontSize: global_FontSize_md.value,
    minHeight: '60px',
    minWidth: '175px',
    width: '50%',
  },
});
