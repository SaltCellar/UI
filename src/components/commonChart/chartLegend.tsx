import { css } from '@patternfly/react-styles';
import { StyleSheet } from '@patternfly/react-styles';
import React from 'react';

const styles = StyleSheet.create({
  legend: {
    display: 'flex',
    alignItems: 'center',
  },
});

interface ChartLegendProps {
  children: React.ReactNode;
  style?: any;
}

const ChartLegend: React.SFC<ChartLegendProps> = ({ children, style }) => {
  const styling =
    Boolean(style) && Boolean(style.legend) ? style.legend : styles.legend;
  return <div className={css(styling)}>{children}</div>;
};

export { ChartLegend, ChartLegendProps };
