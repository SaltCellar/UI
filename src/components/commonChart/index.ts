export { ChartLegend, ChartLegendProps } from './chartLegend';
export { ChartLegendItem, ChartLegendItemProps } from './chartLegendItem';
export { ChartTitle, ChartTitleProps } from './chartTitle';
export { ChartDatum } from './chartUtils';
