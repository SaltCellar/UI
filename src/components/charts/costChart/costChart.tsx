import {
  Chart,
  ChartArea,
  ChartAxis,
  ChartContainer,
  ChartLegend,
  ChartTheme,
  ChartVoronoiContainer,
} from '@patternfly/react-charts';
import { css } from '@patternfly/react-styles';
import { ChartLabelTooltip } from 'components/charts/chartLabelTooltip';
import {
  ChartDatum,
  getDateRange,
  getDateRangeString,
  getMaxValue,
  getMonthRangeString,
  getTooltipContent,
  getTooltipLabel,
} from 'components/charts/commonChart/chartUtils';
import getDate from 'date-fns/get_date';
import React from 'react';
import { FormatOptions, ValueFormatter } from 'utils/formatValue';
import { DomainTuple, VictoryStyleInterface } from 'victory';
import { chartStyles, styles } from './costChart.styles';

interface CostChartProps {
  currentInfrastructureData?: any;
  currentUsageData: any;
  formatDatumValue?: ValueFormatter;
  formatDatumOptions?: FormatOptions;
  height?: number;
  previousInfrastructureData?: any;
  previousUsageData?: any;
  title?: string;
}

interface CostChartDatum {
  data?: any;
  show?: boolean;
  style?: VictoryStyleInterface;
}

interface UsageNameDatum {
  name?: string;
}

interface UsageLegendDatum {
  colorScale?: string[];
  data?: UsageNameDatum[];
  gutter?: number;
  onClick?: (props) => void;
  title?: string;
}

interface Data {
  charts?: CostChartDatum[];
  legend?: UsageLegendDatum;
}

interface State {
  datum?: {
    current?: Data;
    previous?: Data;
  };
  width: number;
}

class CostChart extends React.Component<CostChartProps, State> {
  private containerRef = React.createRef<HTMLDivElement>();
  public state: State = {
    width: 0,
  };

  public componentDidMount() {
    setTimeout(() => {
      if (this.containerRef.current) {
        this.setState({ width: this.containerRef.current.clientWidth });
      }
      window.addEventListener('resize', this.handleResize);
    });
    this.initDatum();
  }

  public componentDidUpdate(prevProps: CostChartProps) {
    if (
      prevProps.currentInfrastructureData !==
        this.props.currentInfrastructureData ||
      prevProps.currentUsageData !== this.props.currentUsageData ||
      prevProps.previousInfrastructureData !==
        this.props.previousInfrastructureData ||
      prevProps.previousUsageData !== this.props.previousUsageData
    ) {
      this.initDatum();
    }
  }

  public componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  private initDatum = () => {
    const {
      currentInfrastructureData,
      currentUsageData,
      previousInfrastructureData,
      previousUsageData,
      title,
    } = this.props;

    // Show all legends, regardless of length -- https://github.com/project-koku/koku-ui/issues/248
    const previousLegendData = [];
    if (previousUsageData) {
      const [start] = getMonthRangeString(previousUsageData, 'chart.cost');
      previousLegendData.push({
        name: start,
        symbol: {
          type: 'minus',
        },
        tooltip: getDateRangeString(previousUsageData, true, true),
      });
    }
    if (previousInfrastructureData) {
      const [start] = getMonthRangeString(
        previousInfrastructureData,
        'chart.infrastructure_cost'
      );
      previousLegendData.push({
        name: start,
        symbol: {
          type: 'dash',
        },
        tooltip: getDateRangeString(previousInfrastructureData, true, true),
      });
    }

    const previous = {
      charts: [
        {
          data: previousUsageData,
          show: true,
          style: chartStyles.previousUsageData,
        },
        {
          data: previousInfrastructureData,
          show: true,
          style: chartStyles.previousInfrastructureData,
        },
      ],
      legend: {
        colorScale: chartStyles.previousColorScale,
        data: previousLegendData,
        onClick: this.handlePreviousLegendClick,
        title,
      },
    };

    const currentLegendData = [];
    if (currentUsageData) {
      const [start] = getMonthRangeString(currentUsageData, 'chart.cost');
      currentLegendData.push({
        name: start,
        symbol: {
          type: 'minus',
        },
        tooltip: getDateRangeString(currentUsageData, true, false),
      });
    }
    if (currentInfrastructureData) {
      const [start] = getMonthRangeString(
        currentInfrastructureData,
        'chart.infrastructure_cost'
      );
      currentLegendData.push({
        name: start,
        symbol: {
          type: 'dash',
        },
        tooltip: getDateRangeString(currentInfrastructureData, true, false),
      });
    }
    const current = {
      charts: [
        {
          data: currentUsageData,
          show: true,
          style: chartStyles.currentUsageData,
        },
        {
          data: currentInfrastructureData,
          show: true,
          style: chartStyles.currentInfrastructureData,
        },
      ],
      legend: {
        colorScale: chartStyles.currentColorScale,
        data: currentLegendData,
        gutter: 55,
        onClick: this.handleCurrentLegendClick,
        title,
      },
    };

    this.setState({
      datum: {
        previous,
        current,
      },
    });
  };

  private handleCurrentLegendClick = props => {
    const { datum } = this.state;
    const newDatum = { ...datum };

    if (props.index >= 0 && newDatum.current.charts.length) {
      newDatum.current.charts[props.index].show = !newDatum.current.charts[
        props.index
      ].show;
      this.setState({ datum: newDatum });
    }
  };

  private handleResize = () => {
    if (this.containerRef.current) {
      this.setState({ width: this.containerRef.current.clientWidth });
    }
  };

  private handlePreviousLegendClick = props => {
    const { datum } = this.state;
    const newDatum = { ...datum };

    if (props.index >= 0 && newDatum.previous.charts.length) {
      newDatum.previous.charts[props.index].show = !newDatum.previous.charts[
        props.index
      ].show;
      this.setState({ datum: newDatum });
    }
  };

  private getChart = (datum: CostChartDatum, index: number) => {
    if (datum.data && datum.data.length && datum.show) {
      return (
        <ChartArea
          data={datum.data}
          key={`usage-chart-${index}`}
          style={datum.style}
        />
      );
    } else {
      return null;
    }
  };

  private getDomain() {
    const {
      currentInfrastructureData,
      currentUsageData,
      previousInfrastructureData,
      previousUsageData,
    } = this.props;
    const domain: { x: DomainTuple; y?: DomainTuple } = { x: [1, 31] };

    const maxCurrentInfrastructure = currentInfrastructureData
      ? getMaxValue(currentInfrastructureData)
      : 0;
    const maxCurrentUsage = currentUsageData
      ? getMaxValue(currentUsageData)
      : 0;
    const maxPreviousInfrastructure = previousInfrastructureData
      ? getMaxValue(previousInfrastructureData)
      : 0;
    const maxPreviousUsage = previousUsageData
      ? getMaxValue(previousUsageData)
      : 0;
    const maxValue = Math.max(
      maxCurrentInfrastructure,
      maxCurrentUsage,
      maxPreviousInfrastructure,
      maxPreviousUsage
    );
    const max = maxValue > 0 ? Math.ceil(maxValue + maxValue * 0.1) : 0;

    if (max > 0) {
      domain.y = [0, max];
    }
    return domain;
  }

  private getEndDate() {
    const {
      currentInfrastructureData,
      currentUsageData,
      previousInfrastructureData,
      previousUsageData,
    } = this.props;
    const currentInfrastructureDate = currentInfrastructureData
      ? getDate(getDateRange(currentInfrastructureData, true, true)[1])
      : 0;
    const currentUsageDate = currentUsageData
      ? getDate(getDateRange(currentUsageData, true, true)[1])
      : 0;
    const previousInfrastructureDate = previousInfrastructureData
      ? getDate(getDateRange(previousInfrastructureData, true, true)[1])
      : 0;
    const previousUsageDate = previousUsageData
      ? getDate(getDateRange(previousUsageData, true, true)[1])
      : 0;

    return currentInfrastructureDate > 0 ||
      currentUsageDate > 0 ||
      previousInfrastructureDate > 0 ||
      previousUsageDate > 0
      ? Math.max(
          currentInfrastructureDate,
          currentUsageDate,
          previousInfrastructureDate,
          previousUsageDate
        )
      : 31;
  }

  private getLegend = (datum: UsageLegendDatum, width: number) => {
    if (datum && datum.data && datum.data.length) {
      return (
        <ChartLegend
          colorScale={datum.colorScale}
          containerComponent={<ChartContainer responsive={false} />}
          data={datum.data}
          events={[
            {
              target: 'data',
              eventHandlers: {
                onClick: () => {
                  return [
                    {
                      target: 'data',
                      mutation: props => {
                        datum.onClick(props);
                        return null;
                      },
                    },
                  ];
                },
              },
            },
          ]}
          height={25}
          itemsPerRow={1}
          labelComponent={<ChartLabelTooltip content={this.getLegendTooltip} />}
          orientation="horizontal"
          style={chartStyles.legend}
          theme={ChartTheme.light.blue}
        />
      );
    } else {
      return null;
    }
  };

  private getLegendTooltip = (datum: ChartDatum) => {
    return datum.tooltip ? datum.tooltip : '';
  };

  private getTooltipLabel = (datum: ChartDatum) => {
    const { formatDatumValue, formatDatumOptions } = this.props;
    return getTooltipLabel(
      datum,
      getTooltipContent(formatDatumValue),
      formatDatumOptions,
      'date'
    );
  };

  private isCurrentLegendVisible() {
    const { datum } = this.state;

    let result = false;
    if (datum && datum.current.legend && datum.current.legend.data) {
      datum.current.legend.data.forEach(data => {
        if (data.name && data.name.trim() !== '') {
          result = true;
          return;
        }
      });
    }
    return result;
  }

  private isPreviousLegendVisible() {
    const { datum } = this.state;

    let result = false;
    if (datum && datum.previous.legend && datum.previous.legend.data) {
      datum.previous.legend.data.forEach(data => {
        if (data.name && data.name.trim() !== '') {
          result = true;
          return;
        }
      });
    }
    return result;
  }

  public render() {
    const { height } = this.props;
    const { datum, width } = this.state;

    const container = (
      <ChartVoronoiContainer
        labels={this.getTooltipLabel}
        voronoiDimension="x"
      />
    );
    const legendWidth =
      chartStyles.legend.minWidth > width / 2
        ? chartStyles.legend.minWidth
        : width / 2;
    const domain = this.getDomain();

    const endDate = this.getEndDate();
    const midDate = Math.floor(endDate / 2);

    return (
      <div className={css(styles.chartContainer)} ref={this.containerRef}>
        <Chart
          containerComponent={container}
          domain={domain}
          height={height}
          width={width}
        >
          {Boolean(datum && datum.previous) &&
            datum.previous.charts.map((chart, index) => {
              return this.getChart(chart, index);
            })}
          {Boolean(datum && datum.current) &&
            datum.current.charts.map((chart, index) => {
              return this.getChart(chart, index);
            })}
          <ChartAxis
            style={chartStyles.xAxis}
            tickValues={[1, midDate, endDate]}
          />
          <ChartAxis dependentAxis style={chartStyles.yAxis} />
        </Chart>
        {Boolean(this.isPreviousLegendVisible()) && (
          <div className={css(styles.chartTitle)}>
            {datum.previous.legend.title}
          </div>
        )}
        {Boolean(
          this.isCurrentLegendVisible() && !this.isPreviousLegendVisible()
        ) && (
          <div className={css(styles.chartTitle)}>
            {datum.current.legend.title}
          </div>
        )}
        {Boolean(this.isPreviousLegendVisible()) && (
          <div className={css(styles.legend)}>
            {this.getLegend(datum.previous.legend, legendWidth)}
          </div>
        )}
        {Boolean(this.isCurrentLegendVisible()) && (
          <div className={css(styles.legend)}>
            {this.getLegend(datum.current.legend, legendWidth)}
          </div>
        )}
      </div>
    );
  }
}

export { CostChart, CostChartProps };
