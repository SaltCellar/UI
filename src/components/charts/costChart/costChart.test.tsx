jest.mock('date-fns/format');

import { Chart, ChartArea } from '@patternfly/react-charts';
import { AwsReport, AwsReportData } from 'api/awsReports';
import * as utils from 'components/charts/commonChart/chartUtils';
import formatDate from 'date-fns/format';
import { shallow } from 'enzyme';
import React from 'react';
import { CostChart, CostChartProps } from './costChart';

const currentMonthReport: AwsReport = createReport('1-15-18');
const previousMonthReport: AwsReport = createReport('12-15-17');

const currentUsageData = utils.transformAwsReport(
  currentMonthReport,
  utils.ChartType.daily
);
const previousUsageData = utils.transformAwsReport(
  previousMonthReport,
  utils.ChartType.daily
);

jest.spyOn(utils, 'getTooltipLabel');

const getTooltipLabel = utils.getTooltipLabel as jest.Mock;

const props: CostChartProps = {
  title: 'Infrastructure Title',
  height: 100,
  formatDatumValue: jest.fn(),
  currentUsageData,
  previousUsageData,
  formatDatumOptions: {},
};

test('reports are formatted to datums', () => {
  const view = shallow(<CostChart {...props} />);
  const charts = view.find(ChartArea);
  expect(charts.length).toBe(2);
  expect(charts.at(0).prop('data')).toMatchSnapshot('previous month data');
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('null previous and current reports are handled', () => {
  const view = shallow(
    <CostChart {...props} currentUsageData={null} previousUsageData={null} />
  );
  const charts = view.find(ChartArea);
  expect(charts.length).toBe(0);
});

test('height from props is used', () => {
  const view = shallow(<CostChart {...props} />);
  expect(view.find(Chart).prop('height')).toBe(props.height);
});

test('labels formats with datum and value formatted from props', () => {
  const tooltipFormatMock = jest.spyOn(utils, 'getTooltipContent');
  const formatLabel = jest.fn();
  tooltipFormatMock.mockImplementation(() => formatLabel);

  const view = shallow(<CostChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '1-1-1',
    units: 'units',
  };
  const group = view.find(Chart);
  group.props().containerComponent.props.labels(datum);
  expect(getTooltipLabel).toBeCalledWith(
    datum,
    formatLabel,
    props.formatDatumOptions,
    'date'
  );
  expect(formatLabel).toBeCalledWith(
    datum.y,
    datum.units,
    props.formatDatumOptions
  );
  expect(formatDate).toBeCalledWith(datum.key, expect.any(String));
  expect(view.find(Chart).prop('height')).toBe(props.height);
});

test('labels ignores datums without a date', () => {
  const view = shallow(<CostChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '',
    units: 'units',
  };
  const group = view.find(Chart);
  const value = group.props().containerComponent.props.labels(datum);
  expect(value).toBe('');
  expect(props.formatDatumValue).not.toBeCalled();
});

test('infrastructure is a running total', () => {
  const multiDayReport: AwsReport = {
    data: [
      createReportDataPoint('1-15-18', 1),
      createReportDataPoint('1-16-18', 2),
    ],
  };
  const multiDaytData = utils.transformAwsReport(
    multiDayReport,
    utils.ChartType.daily
  );
  const view = shallow(
    <CostChart {...props} currentUsageData={multiDaytData} />
  );
  const charts = view.find(ChartArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('infrastructure is a daily value', () => {
  const multiDayReport: AwsReport = {
    data: [
      createReportDataPoint('1-15-18', 1),
      createReportDataPoint('1-16-18', 2),
    ],
  };
  const multiDaytData = utils.transformAwsReport(
    multiDayReport,
    utils.ChartType.daily
  );
  const view = shallow(
    <CostChart {...props} currentUsageData={multiDaytData} />
  );
  const charts = view.find(ChartArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

function createReport(date: string): AwsReport {
  return {
    data: [createReportDataPoint(date)],
  };
}

function createReportDataPoint(date: string, cost = 1): AwsReportData {
  const item = { value: cost, units: 'unit' };
  return {
    date,
    values: [{ date, cost: item }],
  };
}
