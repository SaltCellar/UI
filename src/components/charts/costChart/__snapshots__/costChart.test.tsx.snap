// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infrastructure is a daily value: current month data 1`] = `
Array [
  Object {
    "key": "1-15-18",
    "name": "1-15-18",
    "units": "unit",
    "x": 15,
    "y": 0,
  },
  Object {
    "key": "1-16-18",
    "name": "1-16-18",
    "units": "unit",
    "x": 16,
    "y": 0,
  },
]
`;

exports[`infrastructure is a running total: current month data 1`] = `
Array [
  Object {
    "key": "1-15-18",
    "name": "1-15-18",
    "units": "unit",
    "x": 15,
    "y": 0,
  },
  Object {
    "key": "1-16-18",
    "name": "1-16-18",
    "units": "unit",
    "x": 16,
    "y": 0,
  },
]
`;

exports[`reports are formatted to datums: current month data 1`] = `
Array [
  Object {
    "key": "1-15-18",
    "name": "1-15-18",
    "units": "unit",
    "x": 15,
    "y": 0,
  },
]
`;

exports[`reports are formatted to datums: previous month data 1`] = `
Array [
  Object {
    "key": "12-15-17",
    "name": "12-15-17",
    "units": "unit",
    "x": 15,
    "y": 0,
  },
]
`;
